### PR TITLE
Add multipart/form-data support to uri module

### PR DESCRIPTION
**Pull request by freakinhippie**
Signed-off-by: Joshua Colson <joshua.colson@gmail.com>

##### SUMMARY

Add support for multipart/form-data uploads to the uri module.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

multipart/form-data

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by ahuret** [comment]
Hi ! what is missing in this pr ? @freakinhippie could I help maybe ? Thx

**Comment by freakinhippie** [comment]
I don't think anything is missing. I believe it is just a matter of the ansible project drowning in issues and PRs and there are only so many people to handle them all, but that is just an assumption.

**Review by sivel — CHANGES_REQUESTED** [review]
I have been reviewing this PR, primarily from the perspective of licensing. The code here looks heavily influenced by http://code.activestate.com/recipes/578668-encode-multipart-form-data-for-uploading-files-via/ which is licensed as MIT.  As such, the inclusion of this code in this GPL file is semi-problematic, in that the source and original license are not mentioned.

In addition to this, it appears as though while it was influenced by the above activestate, it seems to be missing additional functionality necessary for this to work as well as the original code, namely things like not escaping quotes in the field and filename.

To extend this further, and I'm not positive how to handle it exactly, but the filenames within the `body`, need to be supported in the `uri` action plugin, so that they can be copied to the target host automatically, just like the `src` parameter is handled.  

I'm also somewhat unconvinced that supplying the configuration in `body` is the correct way to go, I wonder if it's best to add a new discrete argument for supplying the multipart fields.

I also believe this `form_multipart` method should probably be moved to `lib/ansible/module_utils/urls.py`, so that it can be used in a broader scope, instead of only limited to the `uri` module.

I'm in favor of "vendoring" this type of code when it makes sense, but I wonder if utilizing a pre-existing library that provides this functionality is better.  In which case, it would require a `try/except` around the import, and the ability raise a friendly error when the dependency is missing when trying to utilize `multipart/form-data`

In addition, this will require unit and integration tests, as well as a changelog entry.

File: lib/ansible/modules/net_tools/basics/uri.py [hunk 1]
```diff
@@ -479,6 +497,57 @@ def form_urlencoded(body):
...
+                        with open(value['filename'], 'rb') as data_file:
+                            data = data_file.read()
+                    else:
+                        raise IOError('unable to read file "%s"' % value['filename'])
+                result.extend((
+                    '--%s' % boundary,
+                    'Content-Disposition: form-data; name="%s"; filename="%s"' % (field, filename),
+                    'Content-Type: %s' % mimetype,
+                    '',
+                    to_text(data)
```
Comment (cognifloyd) [inline lib/ansible/modules/net_tools/basics/uri.py:542]: If data is binary and not textual, then there is a good chance that passing this through to_text will mangle the binary data.
I just tested this with a wav file, and it seems to mangle it - the textual headers are intact, but the audio content is garbled.
I'll try to think of an alternative for this.

Comment (cognifloyd) [reply lib/ansible/modules/net_tools/basics/uri.py:542]: I hacked this up using `to_bytes` generously. I dropped the `result` list and instead appended bytes to a `b_result` bytes object. Thus, I could merely append the data to the `b_result` without any additional processing.
My hack was very ugly. I can share it if you like, but it is not clean enough to be merged :P

Comment (cognifloyd) [inline lib/ansible/modules/net_tools/basics/uri.py:542]:
```suggestion
                    data
```

File: lib/ansible/modules/net_tools/basics/uri.py [hunk 2]
```diff
@@ -479,6 +497,57 @@ def form_urlencoded(body):
...
+                ))
+        result.extend((
+            '--%s--' % boundary,
+            ''
+        ))
+        return '\r\n'.join(result)
```
Comment (cognifloyd) [inline lib/ansible/modules/net_tools/basics/uri.py:548]:
```suggestion
        return b'\r\n'.join(to_bytes(line, nonstring='passthru') for line in result)
```
(and import `to_bytes`)

**Review by cognifloyd — CHANGES_REQUESTED** [review]
I think these two changes should be enough to preserve binary data as binary.

**Comment by sivel** [comment]
Superseded by https://github.com/ansible/ansible/pull/69376

**Pull request closed without merging**